### DISCUSSION TITLE
Import power data

### DISCUSSION THREAD

**Discussion by PeterV2** [tcgoetz/GarminDB#122] [Q&A]
I got it to work with a virtual machine with Ubuntu, although my knowledge of Linux is almost zero. Thanks for this, very much appreciated.

Is it also possible to import power data for the activities?  I use Stryd for running and a power meter for cycling.

Greetings,

Peter.

**Comment by tcgoetz** [comment]
> Is it also possible to import power data for the activities? I use Stryd for running and a power meter for cycling.

It should be. It will probably require a plugin for the Stryd. If you send me a FIT file with Stryd data, I'll see if I can add one.

**Comment by PeterV2** [comment]
How can I send you fit files?

**Comment by tcgoetz** [comment]
Upload to a file sharing service like Dropbox and send a link.

**Reply by PeterV2** [reply]
I uploaded one run and one bike fitfile:

https://www.dropbox.com/s/40nk7nuqttp17tk/FitFiles.zip?dl=0

**Reply by tcgoetz** [reply]
It says the files were deleted.

![Screen Shot 2021-06-02 at 8 52 24 PM](https://user-images.githubusercontent.com/15090876/120569769-997d0800-c3e4-11eb-9ddd-0fda597ffff7.jpg)

**Comment by PeterV2** [comment]
New link: https://www.dropbox.com/s/h9s3tm39sgnl1if/Fit_files.zip?dl=0

Thanks

**Comment by tcgoetz** [comment]
Is this the data your looking for?

Message: DataMessage(<MessageType.dev_data_id: 207>: [developer_id([invalid] ([255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255])), application_id(bytearray(b'\x18\xfb,\xf0\x1aKC\r\xadf\x98\x8c\x84t!\xf4') ([24, 251, 44, 240, 26, 75, 67, 13, 173, 102, 152, 140, 132, 116, 33, 244])), application_version(83.0), manufacturer([invalid] (65535)), developer_data_index(0)])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Power' ([80, 111, 119, 101, 114, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('Watts' ([87, 97, 116, 116, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(20), developer_data_index(0), field_definition_number(0.0), fit_base_type_id('uint16' (132)), scale([invalid] (255)), offset([invalid] (127)), native_field_num(7)])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Cadence' ([67, 97, 100, 101, 110, 99, 101, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('RPM' ([82, 80, 77, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(20), developer_data_index(0), field_definition_number(2.0), fit_base_type_id('uint8' (2)), scale([invalid] (255)), offset([invalid] (127)), native_field_num(4)])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Ground Time' ([71, 114, 111, 117, 110, 100, 32, 84, 105, 109, 101, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('Milliseconds' ([77, 105, 108, 108, 105, 115, 101, 99, 111, 110, 100, 115, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(20), developer_data_index(0), field_definition_number(3.0), fit_base_type_id('uint16' (132)), scale([invalid] (255)), offset([invalid] (127)), native_field_num(41)])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Vertical Oscillation' ([86, 101, 114, 116, 105, 99, 97, 108, 32, 79, 115, 99, 105, 108, 108, 97, 116, 105, 111, 110, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('Centimeters' ([67, 101, 110, 116, 105, 109, 101, 116, 101, 114, 115, 0, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(20), developer_data_index(0), field_definition_number(4.0), fit_base_type_id('float32' (136)), scale([invalid] (255)), offset([invalid] (127)), native_field_num(39)])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Elevation' ([69, 108, 101, 118, 97, 116, 105, 111, 110, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('Meters' ([77, 101, 116, 101, 114, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(20), developer_data_index(0), field_definition_number(7.0), fit_base_type_id('uint16' (132)), scale([invalid] (255)), offset([invalid] (127)), native_field_num([invalid] (255))])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Form Power' ([70, 111, 114, 109, 32, 80, 111, 119, 101, 114, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('Watts' ([87, 97, 116, 116, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(20), developer_data_index(0), field_definition_number(8.0), fit_base_type_id('uint16' (132)), scale([invalid] (255)), offset([invalid] (127)), native_field_num([invalid] (255))])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Leg Spring Stiffness' ([76, 101, 103, 32, 83, 112, 114, 105, 110, 103, 32, 83, 116, 105, 102, 102, 110, 101, 115, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('kN/m' ([107, 78, 47, 109, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(20), developer_data_index(0), field_definition_number(9.0), fit_base_type_id('float32' (136)), scale([invalid] (255)), offset([invalid] (127)), native_field_num([invalid] (255))])
Message: DataMessage(<MessageType.field_description: 206>: [field_name('Lap Power' ([76, 97, 112, 32, 80, 111, 119, 101, 114, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), units('Watts' ([87, 97, 116, 116, 115, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])), fit_base_unit_id([invalid]), native_message_num(19), developer_data_index(0), field_definition_number(10.0), fit_base_type_id('uint16' (132)), scale([invalid] (255)), offset([invalid] (127)), native_field_num(7)])

**Comment by tcgoetz** [comment]
Which of the Stryd apps did you use for the run and bike fit files? https://apps.garmin.com/en-US/developer/b7fc9884-0f3e-4d29-ba67-401546c01dd8/apps

**Reply by PeterV2** [reply]
Yeah, seems like that's the data.

I am using Garmin Fenix 5X with the app Stryd Zones | Data Field.

https://apps.garmin.com/en-US/apps/18fb2cf0-1a4b-430d-ad66-988c847421f4

**Comment by tcgoetz** [comment]
Please check out and try the develop branch: https://github.com/tcgoetz/GarminDB/commit/67e11403e01ee2b1e5c2cf4f40222f4cd7035402. You will need to `make publish_plugins`.

**Comment by PeterV2** [comment]
Tom,

Thanks for looking into this.
I get an error.



make test_activities_db.TestActivitiesDb.test_fit_file_import
/usr/bin/python3 -m unittest -v test_activities_db.TestActivitiesDb.test_fit_file_import
db params <DbParams() {'db_type': 'sqlite', 'db_path': '/tmp/tmp64k9jm0e/DBs'}
test_fit_file_import (test_activities_db.TestActivitiesDb) ... Processing [<FileType.activity: 4>] FIT data from test_files/fit/activity
  0%|                                                 | 0/1 [00:00<?, ?files/s]Failed to parse test_files/fit/activity/2021-05-17-17-19-10.fit: (raised as a result of Query-invoked autoflush; consider using a session.no_autoflush block if this flush is occurring prematurely)
(builtins.TypeError) SQLite Time type only accepts Python time objects as input.
[SQL: INSERT INTO activity_laps (start_lat, start_long, stop_lat, stop_long, activity_id, lap, start_time, stop_time, elapsed_time, moving_time, distance, cycles, avg_hr, max_hr, avg_rr, max_rr, calories, avg_cadence, max_cadence, avg_speed, max_speed, ascent, descent, max_temperature, min_temperature, avg_temperature) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]
[parameters: [{'start_lat': 51.953862151131034, 'elapsed_time': 220.0, 'start_time': datetime.datetime(2021, 5, 17, 17, 19, 10), 'stop_time': datetime.datetime(202 ... (391 characters truncated) ... ong': 4.53893962316215, 'avg_hr': 139.0, 'max_hr': 157.0, 'max_temperature': 80.6, 'lap': 0, 'avg_rr': None, 'max_rr': None, 'min_temperature': None}]]
100%|█████████████████████████████████████████| 1/1 [00:16<00:00, 16.42s/files]
/home/peter/git/GarminDB/test/test_activities_db.py:92: ResourceWarning: unclosed file <_io.TextIOWrapper name='fit_activities_import_cum.txt' mode='w' encoding='UTF-8'>
  self.profile_function('fit_activities_import', self.__fit_file_import)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
/home/peter/git/GarminDB/test/test_activities_db.py:92: ResourceWarning: unclosed file <_io.TextIOWrapper name='fit_activities_import_tot.txt' mode='w' encoding='UTF-8'>
  self.profile_function('fit_activities_import', self.__fit_file_import)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
FAIL

======================================================================
FAIL: test_fit_file_import (test_activities_db.TestActivitiesDb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/peter/git/GarminDB/test/test_activities_db.py", line 118, in test_fit_file_import
    self.fit_file_import()
  File "/home/peter/git/GarminDB/test/test_activities_db.py", line 93, in fit_file_import
    self.check_db_tables_exists(self.test_mon_db, {'device_table' : GarminDB.Device})
  File "/home/peter/git/GarminDB/test/test_db_base.py", line 39, in check_db_tables_exists
    self.assertGreaterEqual(table.row_count(db), min_rows, 'table %s has no data' % table_name)
AssertionError: 0 not greater than or equal to 1 : table device_table has no data

----------------------------------------------------------------------
Ran 1 test in 17.022s

FAILED (failures=1)
make: *** [Makefile:47: test_activities_db.TestActivitiesDb.test_fit_file_import] Error 1


Kind regards,

Peter.

**Reply by tcgoetz** [reply]
Besides the new plugin, there was a fix for the above issue. Does your activity_fit_file_processor.py line 86 look like this:

`                'elapsed_time'                      : message_fields.get('total_elapsed_time'),`

and not this:

`                'elapsed_time'                      : self._get_field_value(message_fields, 'total_elapsed_time'),`

**Comment by PeterV2** [comment]
I had the self.xxx still in activity_fit_file_processor.py
I updated it.
Now it generates a test DB, but I do not see any new (power, etc) data in the database.
There is a ResourceWarning, but I have no idea what it means.

root@peter-virtual-machine:/home/peter/git/GarminDB/test# make test_activities_db.TestActivitiesDb.test_fit_file_import
/usr/bin/python3 -m unittest -v test_activities_db.TestActivitiesDb.test_fit_file_import
db params <DbParams() {'db_type': 'sqlite', 'db_path': '/tmp/tmp9ffpx9ol/DBs'}
test_fit_file_import (test_activities_db.TestActivitiesDb) ... Processing [<FileType.activity: 4>] FIT data from test_files/fit/activity
  0%|                                                                                                                      | 0/1 [00:00<?, ?files/s]writing lap {'activity_id': '2021', 'lap': 0, 'start_time': datetime.datetime(2021, 5, 17, 17, 19, 10), 'stop_time': datetime.datetime(2021, 5, 17, 17, 25, 20), 'elapsed_time': datetime.time(0, 6, 9, 380000), 'moving_time': datetime.time(0, 6, 9, 380000), 'start_lat': 51.953862151131034, 'start_long': 4.53893962316215, 'stop_lat': 51.955963829532266, 'stop_long': 4.527834355831146, 'distance': 0.6213712, 'cycles': 23.0, 'avg_hr': 139.0, 'max_hr': 157.0, 'avg_rr': None, 'max_rr': None, 'calories': 70.0, 'avg_cadence': 80.0, 'max_cadence': 86.0, 'avg_speed': 6.055385751999999, 'max_speed': 7.368467184, 'ascent': 3.2808399, 'descent': 16.4041995, 'max_temperature': 80.6, 'avg_temperature': 66.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 1, 'start_time': datetime.datetime(2021, 5, 17, 17, 25, 21), 'stop_time': datetime.datetime(2021, 5, 17, 17, 30, 48), 'elapsed_time': datetime.time(0, 5, 27, 890000), 'moving_time': datetime.time(0, 5, 27, 890000), 'start_lat': 51.95595586672425, 'start_long': 4.527788842096925, 'stop_lat': 51.94886871613562, 'stop_long': 4.5259541273117065, 'distance': 0.6213712, 'cycles': 445.0, 'avg_hr': 140.0, 'max_hr': 145.0, 'avg_rr': None, 'max_rr': None, 'calories': 67.0, 'avg_cadence': 81.0, 'max_cadence': 82.0, 'avg_speed': 6.8226548000000005, 'max_speed': 7.47136624, 'ascent': 3.2808399, 'descent': 22.9658793, 'max_temperature': 64.4, 'avg_temperature': 60.8} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 2, 'start_time': datetime.datetime(2021, 5, 17, 17, 30, 49), 'stop_time': datetime.datetime(2021, 5, 17, 17, 36, 35), 'elapsed_time': datetime.time(0, 5, 46, 999000), 'moving_time': datetime.time(0, 5, 46, 999000), 'start_lat': 51.94884533062577, 'start_long': 4.525969130918384, 'stop_lat': 51.946311397477984, 'stop_long': 4.515846390277147, 'distance': 0.6213712, 'cycles': 469.0, 'avg_hr': 139.0, 'max_hr': 144.0, 'avg_rr': None, 'max_rr': None, 'calories': 66.0, 'avg_cadence': 81.0, 'max_cadence': 85.0, 'avg_speed': 6.446849552000001, 'max_speed': 7.619004016000001, 'ascent': 0.0, 'descent': 16.4041995, 'max_temperature': 66.2, 'avg_temperature': 62.6} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 3, 'start_time': datetime.datetime(2021, 5, 17, 17, 36, 36), 'stop_time': datetime.datetime(2021, 5, 17, 17, 42, 36), 'elapsed_time': datetime.time(0, 6, 1, 425000), 'moving_time': datetime.time(0, 6, 1, 425000), 'start_lat': 51.94630980491638, 'start_long': 4.5158112701028585, 'stop_lat': 51.94172842428088, 'stop_long': 4.503646111115813, 'distance': 0.6213712, 'cycles': 483.0, 'avg_hr': 138.0, 'max_hr': 142.0, 'avg_rr': None, 'max_rr': None, 'calories': 64.0, 'avg_cadence': 80.0, 'max_cadence': 81.0, 'avg_speed': 6.189601912, 'max_speed': 6.657121536, 'ascent': 0.0, 'descent': 6.5616798, 'max_temperature': 59.0, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 4, 'start_time': datetime.datetime(2021, 5, 17, 17, 42, 37), 'stop_time': datetime.datetime(2021, 5, 17, 17, 51, 36), 'elapsed_time': datetime.time(0, 9, 0, 285000), 'moving_time': datetime.time(0, 7, 55, 693000), 'start_lat': 51.941708475351334, 'start_long': 4.503621635958552, 'stop_lat': 51.93416794762015, 'stop_long': 4.5024538692086935, 'distance': 0.6213712, 'cycles': 329.0, 'avg_hr': 117.0, 'max_hr': 144.0, 'avg_rr': None, 'max_rr': None, 'calories': 57.0, 'avg_cadence': 68.0, 'max_cadence': 84.0, 'avg_speed': 4.702039472, 'max_speed': 6.9501601520000005, 'ascent': 13.1233596, 'descent': 42.650918700000005, 'max_temperature': 59.0, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 5, 'start_time': datetime.datetime(2021, 5, 17, 17, 51, 37), 'stop_time': datetime.datetime(2021, 5, 17, 17, 57, 25), 'elapsed_time': datetime.time(0, 5, 48, 465000), 'moving_time': datetime.time(0, 5, 48, 465000), 'start_lat': 51.93415361456573, 'start_long': 4.5024750754237175, 'stop_lat': 51.94107321090996, 'stop_long': 4.509316133335233, 'distance': 0.6213712, 'cycles': 467.0, 'avg_hr': 135.0, 'max_hr': 144.0, 'avg_rr': None, 'max_rr': None, 'calories': 61.0, 'avg_cadence': 80.0, 'max_cadence': 83.0, 'avg_speed': 6.420006320000001, 'max_speed': 7.075428568, 'ascent': 36.0892389, 'descent': 9.8425197, 'max_temperature': 59.0, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 6, 'start_time': datetime.datetime(2021, 5, 17, 17, 57, 26), 'stop_time': datetime.datetime(2021, 5, 17, 18, 3, 3), 'elapsed_time': datetime.time(0, 5, 38, 756000), 'moving_time': datetime.time(0, 5, 38, 756000), 'start_lat': 51.94108779542148, 'start_long': 4.5093518402427435, 'stop_lat': 51.940462505444884, 'stop_long': 4.521803576499224, 'distance': 0.6213712, 'cycles': 450.0, 'avg_hr': 142.0, 'max_hr': 146.0, 'avg_rr': None, 'max_rr': None, 'calories': 67.0, 'avg_cadence': 80.0, 'max_cadence': 81.0, 'avg_speed': 6.603435072, 'max_speed': 7.263331192, 'ascent': 3.2808399, 'descent': 13.1233596, 'max_temperature': 62.6, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 7, 'start_time': datetime.datetime(2021, 5, 17, 18, 3, 4), 'stop_time': datetime.datetime(2021, 5, 17, 18, 8, 35), 'elapsed_time': datetime.time(0, 5, 31, 37000), 'moving_time': datetime.time(0, 5, 31, 37000), 'start_lat': 51.940452698618174, 'start_long': 4.521842049434781, 'stop_lat': 51.93382169120014, 'stop_long': 4.5227946527302265, 'distance': 0.6213712, 'cycles': 443.0, 'avg_hr': 143.0, 'max_hr': 148.0, 'avg_rr': None, 'max_rr': None, 'calories': 67.0, 'avg_cadence': 80.0, 'max_cadence': 81.0, 'avg_speed': 6.757783656, 'max_speed': 7.3058329760000005, 'ascent': 0.0, 'descent': 3.2808399, 'max_temperature': 62.6, 'avg_temperature': 59.0} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 8, 'start_time': datetime.datetime(2021, 5, 17, 18, 8, 36), 'stop_time': datetime.datetime(2021, 5, 17, 18, 14, 14), 'elapsed_time': datetime.time(0, 5, 39, 636000), 'moving_time': datetime.time(0, 5, 39, 636000), 'start_lat': 51.93379386328161, 'start_long': 4.522776966914535, 'stop_lat': 51.928739324212074, 'stop_long': 4.515839600935578, 'distance': 0.6213712, 'cycles': 449.0, 'avg_hr': 142.0, 'max_hr': 145.0, 'avg_rr': None, 'max_rr': None, 'calories': 67.0, 'avg_cadence': 79.0, 'max_cadence': 81.0, 'avg_speed': 6.585539584, 'max_speed': 7.18056456, 'ascent': 3.2808399, 'descent': 0.0, 'max_temperature': 64.4, 'avg_temperature': 59.0} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 9, 'start_time': datetime.datetime(2021, 5, 17, 18, 14, 15), 'stop_time': datetime.datetime(2021, 5, 17, 18, 21, 50), 'elapsed_time': datetime.time(0, 7, 36, 24000), 'moving_time': datetime.time(0, 7, 36, 24000), 'start_lat': 51.92875633947551, 'start_long': 4.515812695026398, 'stop_lat': 51.93354190327227, 'stop_long': 4.503781897947192, 'distance': 0.6213712, 'cycles': 517.0, 'avg_hr': 123.0, 'max_hr': 146.0, 'avg_rr': None, 'max_rr': None, 'calories': 65.0, 'avg_cadence': 68.0, 'max_cadence': 81.0, 'avg_speed': 4.905600648, 'max_speed': 7.032926784000001, 'ascent': 22.9658793, 'descent': 6.5616798, 'max_temperature': 62.6, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 10, 'start_time': datetime.datetime(2021, 5, 17, 18, 21, 51), 'stop_time': datetime.datetime(2021, 5, 17, 18, 27, 36), 'elapsed_time': datetime.time(0, 5, 45, 339000), 'moving_time': datetime.time(0, 5, 45, 339000), 'start_lat': 51.933549111709, 'start_long': 4.503755494952202, 'stop_lat': 51.94091713987291, 'stop_long': 4.503963869065046, 'distance': 0.6213712, 'cycles': 466.0, 'avg_hr': 134.0, 'max_hr': 141.0, 'avg_rr': None, 'max_rr': None, 'calories': 50.0, 'avg_cadence': 80.0, 'max_cadence': 84.0, 'avg_speed': 6.478166656, 'max_speed': 7.032926784000001, 'ascent': 6.5616798, 'descent': 16.4041995, 'max_temperature': 68.0, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 11, 'start_time': datetime.datetime(2021, 5, 17, 18, 27, 37), 'stop_time': datetime.datetime(2021, 5, 17, 18, 33, 10), 'elapsed_time': datetime.time(0, 5, 34, 539000), 'moving_time': datetime.time(0, 5, 34, 539000), 'start_lat': 51.94093675352633, 'start_long': 4.503984153270721, 'stop_lat': 51.94590797647834, 'stop_long': 4.515554700046778, 'distance': 0.6213712, 'cycles': 445.0, 'avg_hr': 138.0, 'max_hr': 142.0, 'avg_rr': None, 'max_rr': None, 'calories': 56.0, 'avg_cadence': 80.0, 'max_cadence': 82.0, 'avg_speed': 6.686201704, 'max_speed': 7.095560992, 'ascent': 0.0, 'descent': 16.4041995, 'max_temperature': 60.8, 'avg_temperature': 55.400000000000006} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 12, 'start_time': datetime.datetime(2021, 5, 17, 18, 33, 11), 'stop_time': datetime.datetime(2021, 5, 17, 18, 39, 4), 'elapsed_time': datetime.time(0, 5, 53, 911000), 'moving_time': datetime.time(0, 5, 35, 618000), 'start_lat': 51.94592172279954, 'start_long': 4.515595603734255, 'stop_lat': 51.94662882015109, 'stop_long': 4.526979066431522, 'distance': 0.6213712, 'cycles': 455.0, 'avg_hr': 137.0, 'max_hr': 141.0, 'avg_rr': None, 'max_rr': None, 'calories': 53.0, 'avg_cadence': 80.0, 'max_cadence': 82.0, 'avg_speed': 6.66606928, 'max_speed': 7.2006969839999995, 'ascent': 0.0, 'descent': 9.8425197, 'max_temperature': 55.400000000000006, 'avg_temperature': 53.6} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 13, 'start_time': datetime.datetime(2021, 5, 17, 18, 39, 5), 'stop_time': datetime.datetime(2021, 5, 17, 18, 44, 38), 'elapsed_time': datetime.time(0, 5, 33, 814000), 'moving_time': datetime.time(0, 5, 33, 814000), 'start_lat': 51.946656899526715, 'start_long': 4.526967415586114, 'stop_lat': 51.95529612712562, 'stop_long': 4.52450355514884, 'distance': 0.6213712, 'cycles': 449.0, 'avg_hr': 139.0, 'max_hr': 144.0, 'avg_rr': None, 'max_rr': None, 'calories': 55.0, 'avg_cadence': 81.0, 'max_cadence': 82.0, 'avg_speed': 6.701860256, 'max_speed': 7.47136624, 'ascent': 9.8425197, 'descent': 0.0, 'max_temperature': 57.2, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 14, 'start_time': datetime.datetime(2021, 5, 17, 18, 44, 39), 'stop_time': datetime.datetime(2021, 5, 17, 18, 50, 12), 'elapsed_time': datetime.time(0, 5, 34, 648000), 'moving_time': datetime.time(0, 5, 34, 648000), 'start_lat': 51.95532345212996, 'start_long': 4.524489725008607, 'stop_lat': 51.955860732123256, 'stop_long': 4.536407869309187, 'distance': 0.6213712, 'cycles': 447.0, 'avg_hr': 137.0, 'max_hr': 140.0, 'avg_rr': None, 'max_rr': None, 'calories': 51.0, 'avg_cadence': 80.0, 'max_cadence': 82.0, 'avg_speed': 6.683964768, 'max_speed': 7.220829408, 'ascent': 6.5616798, 'descent': 3.2808399, 'max_temperature': 57.2, 'avg_temperature': 57.2} for test_files/fit/activity/2021-05-17-17-19-10.fit
writing lap {'activity_id': '2021', 'lap': 15, 'start_time': datetime.datetime(2021, 5, 17, 18, 50, 13), 'stop_time': datetime.datetime(2021, 5, 17, 18, 54, 20), 'elapsed_time': datetime.time(0, 1, 52, 16000), 'moving_time': datetime.time(0, 1, 52, 16000), 'start_lat': 51.95583634078503, 'start_long': 4.53641390427947, 'stop_lat': 51.95389718748629, 'stop_long': 4.53890785574913, 'distance': 0.19793779576, 'cycles': 148.0, 'avg_hr': 135.0, 'max_hr': 137.0, 'avg_rr': None, 'max_rr': None, 'calories': 16.0, 'avg_cadence': 79.0, 'max_cadence': 81.0, 'avg_speed': 6.3618459839999995, 'max_speed': 6.992661936, 'ascent': 0.0, 'descent': 0.0, 'max_temperature': 55.400000000000006, 'avg_temperature': 55.400000000000006} for test_files/fit/activity/2021-05-17-17-19-10.fit
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:20<00:00, 20.10s/files]
/home/peter/git/GarminDB/test/test_activities_db.py:92: ResourceWarning: unclosed file <_io.TextIOWrapper name='fit_activities_import_cum.txt' mode='w' encoding='UTF-8'>
  self.profile_function('fit_activities_import', self.__fit_file_import)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
/home/peter/git/GarminDB/test/test_activities_db.py:92: ResourceWarning: unclosed file <_io.TextIOWrapper name='fit_activities_import_tot.txt' mode='w' encoding='UTF-8'>
  self.profile_function('fit_activities_import', self.__fit_file_import)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
ok

----------------------------------------------------------------------
Ran 1 test in 20.841s

OK

**Reply by tcgoetz** [reply]
Did you do `make update` to update all of you repos? Do you have a file stryd_zone_plugin.py in your Plugins directory? Do you have these lines in your activities_db.log after running the test?
```
Loaded plugin stryd_zones: <class 'types.stryd_zones'>
Instantiated plugin stryd_zones: <types.stryd_zones object at 0x106e7d3a0>
```

**Comment by PeterV2** [comment]
I have no idea what you mean.
If I run make uopdate I get:

root@peter-virtual-machine:/home/peter/git/GarminDB# make update
git submodule init
git submodule update
git pull --rebase
error: cannot pull with rebase: You have unstaged changes.
error: please commit or stash them.
make: *** [Makefile:48: update] Error 128

Sorry I don't get it, I don't have knowledge about this.

**Reply by tcgoetz** [reply]
> error: cannot pull with rebase: You have unstaged changes.

This error means you have changes in your tree. You need to have no changes in your tree to do `make updates. Remove the changes in your tree and do a `make updates` And then make publish_plugins

**Comment by PeterV2** [comment]
I cannot get it to work. I will leave it. Thanks for all the effort.
